blocks/datasets/__init__.py

```diff
@@ -130,6 +143,37 @@ def get_default_stream(self):
         return DataStream(self, iteration_scheme=self.default_scheme)
 
 
+def lazy_properties(*lazy_properties):
+    def lazy_property_factory(lazy_property):
+        def lazy_property_getter(self):
+            if not hasattr(self, '_' + lazy_property):
+                self.load()
+            if not hasattr(self, '_' + lazy_property):
+                raise ValueError("{} wasn't loaded".format(lazy_property))
+            return getattr(self, '_' + lazy_property)
+
+        def lazy_property_setter(self, value):
+            setattr(self, '_' + lazy_property, value)
+
+        return lazy_property_getter, lazy_property_setter
+
+    def wrap_dataset(dataset):
+        for lazy_property in lazy_properties:
+            setattr(dataset, lazy_property,
+                    property(*lazy_property_factory(lazy_property)))
+        if not hasattr(dataset, '__getstate__'):
+            def __getstate__(self):
+                for lazy_property in lazy_properties:
+                    attr = getattr(self, '_' + lazy_property)
+                    if isinstance(attr, collections.Iterator):
+                        raise ValueError("Iterators can't be lazy loaded")
+                    delattr(self, '_' + lazy_property)
```

Comment:
Wouldn't it be better to create a shallow copy of `__dict__` and change it? I do not like that `__getstate__` changes the object.